src/include/kdb.h.in

```diff
@@ -134,80 +125,61 @@ int kdbSet (KDB *handle, KeySet *returned,
-
-int keyClear (Key *key);
-int keyDel (Key *key);
-
-uint16_t keyIncRef (Key *key);
-uint16_t keyDecRef (Key *key);
-uint16_t keyGetRef (const Key *key);
+ElektraKey * elektraKeyNew (const char * keyname, ...) ELEKTRA_SENTINEL;
+ElektraKey * elektraKeyVNew (const char * keyname, va_list ap);
 
-/* Meta Info */
-int keyRewindMeta (Key *key);
-const Key *keyNextMeta (Key *key);
-const Key *keyCurrentMeta (const Key *key);
+ElektraKey * elektraKeyCopy (ElektraKey * dest, const ElektraKey * source, elektraCopyFlags flags);
 
-int keyCopyMeta (Key *dest, const Key *source, const char *metaName);
-int keyCopyAllMeta (Key *dest, const Key *source);
+int elektraKeyClear (ElektraKey * key);
+int elektraKeyDel (ElektraKey * key);
 
-const Key *keyGetMeta (const Key *key, const char* metaName);
-ssize_t    keySetMeta (Key *key, const char* metaName,
-	const char *newMetaString);
-KeySet * keyMeta (Key * key);
+uint16_t elektraKeyIncRef (ElektraKey * key);
+uint16_t elektraKeyDecRef (ElektraKey * key);
+uint16_t elektraKeyGetRef (const ElektraKey * key);
 
-/* Methods for Making Tests */
-int keyCmp (const Key *k1, const Key *k2);
+/* Metadata */
+ElektraKeySet * elektraKeyMeta (ElektraKey * key);
+int elektraKeySetMeta (ElektraKey * key, ElektraKeySet * meta);
 
-int keyNeedSync (const Key *key);
+/* Relations */
+int elektraKeyNameCmp (const ElektraKey * k1, const ElektraKey * k2);
 
-int keyIsBelow (const Key *key, const Key *check);
-int keyIsBelowOrSame (const Key *key, const Key *check);
-int keyIsDirectlyBelow (const Key *key, const Key *check);
-
-int keyIsBinary (const Key *key);
-int keyIsString (const Key *key);
+int elektraKeyIsBelow (const ElektraKey * key, const ElektraKey * check);
+int elektraKeyIsBelowOrSame (const ElektraKey * key, const ElektraKey * check);
+int elektraKeyIsDirectlyBelow (const ElektraKey * key, const ElektraKey * check);
 
 /* Name Manipulation Methods */
-const char *keyName (const Key *key);
-ssize_t keyGetNameSize (const Key *key);
-ssize_t keyGetName (const Key *key, char *returnedName, size_t maxSize);
-
-ssize_t keySetName (Key *key, const char *newname);
-ssize_t keyAddName (Key *key, const char *addName);
+const char * elektraKeyName (const ElektraKey * key);
+ssize_t elektraKeyNameSize (const ElektraKey * key);
 
-const void *keyUnescapedName (const Key *key);
-ssize_t keyGetUnescapedNameSize (const Key *key);
+ssize_t elektraKeySetName (ElektraKey * key, const char * newname);
+ssize_t elektraKeyAddName (ElektraKey * key, const char * addName);
 
-const char *keyBaseName (const Key *key);
-ssize_t keyGetBaseNameSize (const Key *key);
-ssize_t keyGetBaseName (const Key *key, char *returned, size_t maxSize);
+const char * elektraKeyEscapedName (const ElektraKey * key);
 
-ssize_t keySetBaseName (Key *key,const char *baseName);
-ssize_t keyAddBaseName (Key *key,const char *baseName);
+const char * elektraKeyBaseName (const ElektraKey * key);
+ssize_t elektraKeyBaseNameSize (const ElektraKey * key);
```

Comment:
Not really API, but the base name size should be stored in the struct as an optimization.

Reply:
No, without extensive benchmarks that this is really something that pays off (which I doubt) this is a bad idea (classical premature optimization).

Reply:
I assumed that we would remove fields from `struct _Key`. If the space gained there is used for the base name size, we don't loose anything in terms of memory. In terms of performance, we may not gain much, but storing the size is definitely not slower.

> extensive benchmarks [...] something that pays off (which I doubt)

It is very hard to say, how often the base name size is needed by real world use cases. This makes constructing valid benchmarks very hard.

`elektraKeyBaseNameSize` is probably almost never called directly. However, if the size were know, it could be used in `elektraKeyBaseName` and `elektraKeySetBaseName` (instead of iterating backwards). Those functions are definitely used more (how much is uncertain).

One thing however is certain: There is a (albeit small) performance gain when switching from:

```c
const char * baseName = key->ukey + key->keyUSize - 2;
while (*baseName != '\0')
{
	--baseName;
}
return baseName + 1;
```
to just
```c
return key->ukey + key->keyUSize - key->baseNameSize;
```

The remaining question is: Is this snippet (or the equivalent in `elektraKeySetBaseName`) called often enough that this difference adds up to something significant?

However, once again, this question is only really relevant, if there is a possible downside to adding `key->baseNameSize`.

Reply:
The problem of invalid benchmarks is much smaller than of invalid assumptions in the guesswork of what might be faster.

I don't see a need to have the size stored and without conclusive data that shows the need we shouldn't add it as we already know we are too memory-intensive. So actually we also shouldn't provide API that returns it.

Reply:
> IMO making reasonable assumptions is much better than making confident mistakes based on artificial benchmarks. If your benchmarks don't reflect the real world, basing decisions off of those benchmarks might make things a lot worse.

Making things more objective (or said differently: working more scientific) per se cannot "make things a lot worse". If we get too confident about data that actually does not provide this kind of confidence (i.e. always), we have a similar but not identical problem as when we simply confidently claim something without any backup (i.e. pure opinion).

> Of course the best middle ground would be to make assumptions about the real world and build benchmarks based on that. However, benchmarks for the this example are actually really hard to design. You need to know how often the basename is read/updated to know how much speedup/slowdown would be noticeable. But you also need to know how big basenames are, since the current implementation has runtime O(size of basename).

Kurt Micheli made some very good work in the direction of generating KeySets with different basename lengths. If your benchmark can show improvements across many different basename sizes/KeySet shapes, the "realism" of how the KeySet looks like gets less important. See benchmarks/opmphm.c

Reply:
I don't think this discussion will lead anywhere. It is clear to me that you are absolutely insistent on having benchmarks before changing anything, even if the change can be clearly doesn't impact memory use (same `struct` size) and is from a `O(size of basename)` to a `O(1)` runtime.

I assume this is because, there is the option to reduce the memory footprint (by removing the second size field from the struct) and keeping the current performance. But that means we need a benchmark must show whether the increased performance of the `O(1)` implementation is worth the increased memory footprint.

Such a trade-off can, however, only be evaluated with real-world data. Without the real-world data you can only try all possible combinations of basename size, and number of calls to the various `elektraKey*BaseName*` functions. This will generate huge amounts of data and then what? It's still a trade-off between memory and runtime you can't decided that without real world data.

> See benchmarks/opmphm.c

That is a totally different problem. That benchmark compares the runtime of two implementations. We could do this here too, but I don't see the point since one implementation is `O(1)` (actually constant in all cases and not just mostly constant like a hashmap) and the other is not.

---

In summary, I think we can leave it for now and not store the basename size. In future we can revisit this and maybe then we have some real world data.